vhdbuilder/packer/install-dependencies.sh

```diff
@@ -351,28 +351,104 @@ INSTALLED_RUNC_VERSION=$(runc --version | head -n1 | sed 's/runc version //')
-  ctr -n k8s.io image pull $NVIDIA_DRIVER_IMAGE:$NVIDIA_DRIVER_IMAGE_TAG
-  if grep -q "fullgpu" <<< "$FEATURE_FLAGS"; then
-    bash -c "$CTR_GPU_INSTALL_CMD $NVIDIA_DRIVER_IMAGE:$NVIDIA_DRIVER_IMAGE_TAG gpuinstall /entrypoint.sh install"
-    ret=$?
-    if [[ "$ret" != "0" ]]; then
-      echo "Failed to install GPU driver, exiting..."
-      exit $ret
-    fi
+gpu_action=""
+declare -A pulled_gpu_images
+
+# Loop over each GPUContainerImage
+while IFS= read -r gpuImageToBePulled; do
+  # Extract 'cached' field and convert it to lowercase
+  cached=$(echo "${gpuImageToBePulled}" | jq -r '.cached' | tr '[:upper:]' '[:lower:]')
+
+  if [[ "$cached" != "true" ]]; then
+    # Skip images that are not meant to be cached
+    continue
   fi
 
-  cat << EOF >> ${VHD_LOGS_FILEPATH}
-  - nvidia-driver=${NVIDIA_DRIVER_IMAGE_TAG}
-EOF
+  # Extract 'osSelectors' if present
+  osSelectors=$(echo "${gpuImageToBePulled}" | jq -r '.osSelectors // empty')
+
+  shouldPull=0  # Default to not pull
+
+  if [[ -n "$osSelectors" ]]; then
+    # osSelectors is provided; check if current OS and arch match any entry
+    while IFS= read -r selector; do
+      os=$(echo "$selector" | jq -r '.os')
+      arch=$(echo "$selector" | jq -r '.arch')
+
+      if [[ "$os" == "$CURRENT_OS" ]]; then
+        if [[ "$arch" == "$CPU_ARCH" ]]; then
+          shouldPull=1
+          break  # Found a matching selector
+        fi
+      fi
+    done <<< "$(echo "$osSelectors" | jq -c '.[]')"
+  else
+    # No osSelectors provided; decide whether to pull
+    # Assuming we pull the image if no osSelectors are specified
+    shouldPull=1
+  fi
+
+  if [[ "$shouldPull" == "1" ]]; then
+    # Extract image details
+    downloadURL=$(echo "${gpuImageToBePulled}" | jq -r '.downloadURL')
+    imageName=$(echo "$downloadURL" | sed 's/:.*$//')
+
+    # Get the latestVersion
+    latestVersion=$(echo "${gpuImageToBePulled}" | jq -r '.multiArchVersionsV2[0].latestVersion')
+
+    if [[ -z "$latestVersion" || "$latestVersion" == "null" ]]; then
+      echo "Error: latestVersion not found for $imageName"
+      exit 1
+    fi
+
+    fullImage="$imageName:$latestVersion"
+
+    # Pull the image
+    echo "Pulling image: $fullImage"
+    ctr -n k8s.io image pull "$fullImage"
+    if [[ $? -ne 0 ]]; then
+      echo "Failed to pull image: $fullImage"
+      exit 1
+    fi
+
+    # Record the pulled image
+    pulled_gpu_images["$imageName"]="$latestVersion"
+
+    # Set gpu_action if pulling the aks-gpu-cuda image
+    if [[ "$imageName" == "mcr.microsoft.com/aks/aks-gpu-cuda" ]]; then
+      gpu_action="copy"
+
+      # Create necessary directories
+      mkdir -p /opt/{actions,gpu}
+
+      # Check for the "fullgpu" feature flag
+      if grep -q "fullgpu" <<< "$FEATURE_FLAGS"; then
+        echo "Installing GPU driver from image: $fullImage"
+        bash -c "$CTR_GPU_INSTALL_CMD $fullImage gpuinstall /entrypoint.sh install"
+        ret=$?
+        if [[ "$ret" != "0" ]]; then
```

Comment:
Again, avoid more than 2 level nested if. It's hard to keep track which level it is for debugging and readability.